File: js/structural/proxy.js
class DB {
    readData() {
        console.log('reading data from DB')
    }
}

class DBProxy {
    constructor(is_auth) {
        this.is_auth = is_auth
        this.db = new DB()
    }

    readData() {
        if (this.is_auth === true) this.db.readData()
        else console.error('access denied')

    }
}

proxy = new DBProxy(false)
proxy.readData()

proxy2 = new DBProxy(true)
proxy2.readData()

File: python/structural/proxy.py
class DB:
    @staticmethod
    def read_data():
        print('reading data from DB')


class DBProxy:
    def __init__(self, is_auth):
        self.is_auth = is_auth
        self._real_db = DB()

    def read_data(self):
        if self.is_auth:
            self._real_db.read_data()
        else:
            raise TypeError('access denied')


proxy = DBProxy(is_auth=True)
proxy.read_data()

proxy2 = DBProxy(is_auth=False)
proxy2.read_data()
